### PR TITLE
Sched cache resync

### PR DESCRIPTION
**Pull request by anfernee**
**What this PR does / why we need it**:  Scheduler cache comparer
    
    A debug tool that collects resources from api server and compares it
    with the scheduler cache. It currently only compares the node list, but
    it should be easy to extend. The compare is triggered by signal USER2,
    by doing
    
      kill -12 ${SCHED_PID}
    
    The compare result goes to scheduler log.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Towards #60860

**Special notes for your reviewer**: @bsalamat 

**Release note**:
```release-note
None
```

### CODE REVIEW

**Comment by anfernee** [comment]
In order to make it simple, it's signal triggered now. but I am open to suggestions like option, periodical ticker, etc.

File: pkg/scheduler/schedulercache/cache.go [hunk 1]
```diff
@@ -80,6 +80,33 @@ func newSchedulerCache(ttl, period time.Duration, stop <-chan struct{}) *schedul
 	}
 }
 
+// Snapshot takes a snapshot of the current schedulerCache.
```
Comment (anfernee) [inline pkg/scheduler/schedulercache/cache.go]: this has performance impact, now only used for debug purpose.

**Comment by bsalamat** [comment]
/assign
I took a quick look and the direction is good. Please add comparision for Pods as well.

**Comment by anfernee** [comment]
@bsalamat pods added. more comparison can be added too. but at this point, I am not what's more valuable. right now, mismatching pod, node, pdb will be logged.

**Review by bsalamat — COMMENTED** [review]
Thanks, @anfernee!

File: pkg/scheduler/factory/cache_comparer.go [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+		return err
+	}
+
+	snapshot := c.cache.Snapshot()
+
+	if missed, redundant := c.CompareNodes(nodes, snapshot.Nodes); len(missed)+len(redundant) != 0 {
+		glog.Warningf("cache mismatch: missed nodes: %s; redundant nodes: %s", missed, redundant)
+	}
+
+	if missed, redundant := c.ComparePods(pods, snapshot.Nodes); len(missed)+len(redundant) != 0 {
```
Comment (bsalamat) [inline pkg/scheduler/factory/cache_comparer.go]: This part of the PR is not correct and will create false positives. Scheduler keeps assigned pods (pods with PodSpec.NodeName != "") in its cache. Unassigned pods live in scheduler's scheduling queue. For now it is OK to ignore unassigned pods. The code in its current form compares all pods against assigned ones in the scheduler cache. So, if there are pending pods in the cluster, we will see warning messages.

Comment (anfernee) [reply pkg/scheduler/factory/cache_comparer.go]: I am not aware of that.. Let me fix it and rebase this PR.

Comment (anfernee) [reply pkg/scheduler/factory/cache_comparer.go]: Updated @bsalamat

**Comment by bsalamat** [comment]
/lgtm

Thanks, @anfernee!
I think we should probably get this in 1.10 as well. This may help find an important issue that is causing a lot of trouble for many users.

**Comment by anfernee** [comment]
@bsalamat sure. if it's helpful. it touches a lot but overall not risky, it is not affecting the main workflow.

**Comment by bsalamat** [comment]
I don't think it is urgent enough to be merged during the current code freeze. We can wait until after the code freeze and then backport it to 1.10 for the next dot release of 1.10.

**Comment by anfernee** [comment]
i am fine either way.

**Comment by wenjiaswe** [comment]
cc @roycaihw @wenjiaswe

File: pkg/scheduler/schedulercache/cache.go (hunk 1 above)
Comment (resouer) [reply pkg/scheduler/schedulercache/cache.go]: Why don't add this NOTE in the comment here?

File: pkg/scheduler/factory/signal.go [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+limitations under the License.
+*/
+
+package factory
+
+import "syscall"
+
+// compareSignal is the signal to trigger cache compare. For non-windows
+// environment it's SIGUSR2.
+var compareSignal = syscall.SIGUSR2
```
Comment (resouer) [inline pkg/scheduler/factory/signal.go]: I think we will enable this in test-infra, how will test-infra expect to trigger it may I ask?

Comment (anfernee) [reply pkg/scheduler/factory/signal.go]: you mean test this in e2e test? no, it's for whoever interested in triggering the comparison to debug. Nothing would happen except some logs are appended.

File: pkg/scheduler/schedulercache/cache.go (hunk 1 above)
Comment (anfernee) [reply pkg/scheduler/schedulercache/cache.go]: agreed it would be nice to add in comment.

File: pkg/scheduler/factory/signal.go (hunk 3 above)
Comment (resouer) [reply pkg/scheduler/factory/signal.go]: I saw it was mentioned in the issue description: 
> We will probably enable it in our tests and in clusters where customers see stale cache issues

Anyway this question will not block this PR. I will re-lgtm it when CI passed.

Comment (bsalamat) [reply pkg/scheduler/factory/signal.go]: @resouer, Your point is valid. In fact, I also chatted with @anfernee on Slack about periodic triggering of this or enabling it in tests. I believe we should try this feature and make sure that we do not see a lot of false positives. We should also measure its performance overhead in larger tests to ensure that enabling it in tests won't cause timeouts. Once it is refined, we may want to enable it in certain tests.

Comment (resouer) [reply pkg/scheduler/factory/signal.go]: Got it, thanks for explaining.

**Comment by resouer** [comment]
/lgtm

Thanks for quick fix the nits.

**Comment by Random-Liu** [comment]
I guess this changes broke the test `[sig-scheduling] SchedulerPredicates [Serial] validates that taints-tolerations is respected if not matching`.
https://k8s-testgrid.appspot.com/google-gce#gci-gce-serial

The test started failing after this change list https://github.com/kubernetes/kubernetes/compare/3e5596321...025439988. And this PR seems most relevant.

@anfernee @bsalamat

**Comment by resouer** [comment]
And I believe the break is caused by: https://github.com/kubernetes/kubernetes/pull/61089

**Comment by Random-Liu** [comment]
@resouer Thanks for looking into this!

**Pull request merged**